spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -371,12 +376,45 @@ private Schema schemaWithMetadataColumns() {
             .distinct()
             .map(name -> MetadataColumns.metadataColumn(table, name))
             .collect(Collectors.toList());
-    Schema meta = new Schema(fields);
+    // only calculate potential column id collision if meta column was requested
+    Schema meta = fields.isEmpty() ? new Schema(fields) : calculateUnionedSchema(fields);
 
     // schema or rows returned by readers
     return TypeUtil.join(schema, meta);
   }
 
+  private Schema calculateUnionedSchema(List<Types.NestedField> metaColumnFields) {
+    // Calculate used ids by union metadata columns with all base table schemas
+    Set<Integer> currentlyUsedIds =
+        Collections.unmodifiableSet(
+            TypeUtil.indexById(Types.StructType.of(metaColumnFields)).keySet());
+    Set<Integer> allUsedIds =
+        table.schemas().values().stream()
+            .map(currSchema -> TypeUtil.indexById(currSchema.asStruct()).keySet())
+            .reduce(currentlyUsedIds, Sets::union);
+
+    // Calculate ids to reassign
+    Types.StructType partitionType = Partitioning.partitionType(table);
```

Comment:
It's the same but would make more sense to get this from the metadata column.  If the user doesnt pick _partition, then the whole thing is a no-op, by the way

Reply:
yeah that's what I am thinking as well. In `schemaWithMetadataColumns()` I check if metadataFiled is empty, and bypass all deduplication logic is so and we will not hit this function at all, do you think it's sufficient?